### DISCUSSION TITLE
Caching in MPI wrapped distributed LTIModels

### DISCUSSION THREAD

**Discussion by josefinez** [pymor/pymor#1328] [Q&A] [answered]
Using LTIModels in an efficient way is highly dependent on caching. When MPI wrapping a model with `use_with=True`, caching is disabled. 

I'm assuming caching is disabled to ensure that cached solutions are available on all ranks, even if the event loop is not started in `auto_launch`. But if this would be the only reason, the cache could just be cleared.

Is there any other reason why it could be a bad idea to use caching with an mpi distributed (InputOutput) model?

**Comment by sdrave** [comment, marked as answer]
Let's consider the cached `LTIModel.gramian` which returns a `VectorArray`. Even for mpi distributed `Models`, the method is only called on rank 0, so the result, an `MPIVectorArray` is only available there. Caching using a `MemoryRegion` should just work fine, as `MemoryRegion` only keeps a reference to the cached result. In particular, this causes the returned `MPIVectorArray` to stay alive, so `mpi.remove_object` is never called, and all ranks keep their part of the array referenced in `_managed_objects`.

However, for `DiskRegion` the situation is completely different, as this reason pickles the cached values, and `MPIVectorArrays` can't be pickled. One might think about having an mpi-aware `DiskRegion`, but I'm not sure how useful that would be in practice.

So, overall, I would say that one should change `mpi_wrap_model` to only disable caching when disk-based caching is selected. In that case, it probably also makes sense to issue a warning about caching being disabled. An alternative would be to never disable caching and extend `DiskRegion` to ignore `MPIVectorArrays`, again issuing a warning when such an array is encountered.

**Reply by josefinez** [reply]
Thank you for your explanation!
